### PR TITLE
Add New(Map|Program)FromID

### PR DESCRIPTION
**Pull request by florianl**
Signed-off-by: Lehner Florian <dev@der-flo.net>

This PR add the functions to get the FD of a map or program by its ID.

### CODE REVIEW

**Comment by florianl** [comment]
Thanks for your feedback and I've addressed them.

File: syscalls.go [hunk 1]
```diff
@@ -469,26 +469,10 @@ var objNameAllowsDot = internal.FeatureTest("dot in object names", "5.2", func()
...
-	ptr, err := internal.BPF(_MapGetFDByID, unsafe.Pointer(&attr), unsafe.Sizeof(attr))
-	if err != nil {
-		return nil, xerrors.Errorf("can't get fd for map id %d: %w", id, err)
-	}
-	return internal.NewFD(uint32(ptr)), nil
-}
-
-func bpfGetProgramFDByID(id uint32) (*internal.FD, error) {
-	// available from 4.13
+func objGetFDByID(cmd int, fd uint32) (*internal.FD, error) {
```
Comment (tklauser) [inline syscalls.go]: Why is `id` renamed to `fd` here? I think this should remain as `id`.

File: prog.go [hunk 2]
```diff
@@ -419,7 +419,7 @@ func unmarshalProgram(buf []byte) (*Program, error) {
 	// Looking up an entry in a nested map or prog array returns an id,
 	// not an fd.
 	id := internal.NativeEndian.Uint32(buf)
-	fd, err := bpfGetProgramFDByID(id)
+	fd, err := objGetFDByID(_ProgGetFDByID, id)
 	if err != nil {
```
Comment (lmb) [inline prog.go]: You can replace the rest of this function with a call to NewProgramFromFD, same for unmarshalMap.

File: prog_test.go [hunk 3]
```diff
@@ -360,6 +360,41 @@ func TestProgramGetNextID(t *testing.T) {
...
+	var next ProgramID
+
+	if next, err = ProgramGetNextID(ProgramID(0)); err != nil {
+		t.Fatal("Can't get next ID:", err)
+	}
+	if next == ProgramID(0) {
+		t.Fatal("Expected next ID other than 0")
+	}
+
+	if _, err = NewProgramFromID(next); err != nil {
```
Comment (lmb) [inline prog_test.go]: How much work is it to add `Program.ID() (ProgramID, error)` and similar for `Map`? Then this test could be a lot simpler:
```
id, err := prog.ID()
NewProgramFromID(id)
```

File: prog_test.go [hunk 4]
```diff
@@ -360,6 +360,41 @@ func TestProgramGetNextID(t *testing.T) {
...
+		t.Fatalf("Can't get FD for program ID %d: %v", uint32(next), err)
+	}
+
+	// As there can be multiple programs, we use max(uint32) as ProgramID to trigger an expected error.
+	_, err = NewProgramFromID(ProgramID(4294967295))
```
Comment (lmb) [inline prog_test.go]: Why not `math.MaxUint32`?

File: syscalls.go (hunk 1 above)
Comment (lmb) [inline syscalls.go]: Agree with Tobias, also please keep the `bpf` prefix:

```suggestion
func bpfObjGetFDByID(cmd int, id uint32) (*internal.FD, error) {
```

File: prog_test.go (hunk 4 above)
Comment (florianl) [reply prog_test.go]: I didn't want to add another dependency. But you're right, it's easier with `math.MaxUint32`.

File: syscalls.go (hunk 1 above)
Comment (florianl) [reply syscalls.go]: renamed it

File: prog_test.go (hunk 3 above)
Comment (florianl) [reply prog_test.go]: I've added the implementation for `ID()` and updated the tests.

File: syscalls.go (hunk 1 above)
Comment (florianl) [reply syscalls.go]: sorry for the typo. fixed it.

File: prog.go (hunk 2 above)
Comment (florianl) [reply prog.go]: updated this functions as well.

**Pull request merged**